sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java

```diff
@@ -0,0 +1,160 @@
+
+/**
+ * Clients facing {@link FileSystem} utility.
+ */
+public class FileSystems {
+
+  public static final String DEFAULT_SCHEME = "default";
+
+  private static final Pattern URI_SCHEME_PATTERN = Pattern.compile("^[a-zA-Z][-a-zA-Z0-9+.]*$");
+
+  private static final ClassLoader CLASS_LOADER = ReflectHelpers.findClassLoader();
+
+  private static final Map<String, FileSystemRegistrar> REGISTRAR_MAP =
+      Collections.synchronizedMap(new HashMap<String, FileSystemRegistrar>());
+
+  private static final Map<String, PipelineOptions> REGISTRAR_DEFAULT_CONFIG_MAP =
+      Collections.synchronizedMap(new HashMap<String, PipelineOptions>());
+
+  public static void loadFileSystemRegistrars() {
+    REGISTRAR_MAP.clear();
+    Set<FileSystemRegistrar> registrars =
+        Sets.newTreeSet(ReflectHelpers.ObjectsClassComparator.INSTANCE);
+    registrars.addAll(Lists.newArrayList(
+        ServiceLoader.load(FileSystemRegistrar.class, CLASS_LOADER)));
+
+    checkDuplicateScheme(registrars);
+
+    for (FileSystemRegistrar registrar : registrars) {
+      addFileSystemRegistrar(registrar);
+    }
+  }
+
+  /**
+   * Sets the default config for {@link FileSystemRegistrar} that associates with {@code scheme}.
+   *
+   * <p>Syntax: scheme = alpha *( alpha | digit | "+" | "-" | "." )
+   * Upper case letters are treated as the same as lower case letters.
+   */
+  public static void setDefaultConfig(String scheme, PipelineOptions options) {
+    checkArgument(
+        URI_SCHEME_PATTERN.matcher(scheme).matches(),
+        String.format("Scheme: [%s] doesn't match URI syntax: %s",
+            scheme, URI_SCHEME_PATTERN.pattern()));
+    REGISTRAR_DEFAULT_CONFIG_MAP.put(
+        checkNotNull(scheme, "scheme"), checkNotNull(options, "options"));
+  }
+
+  /**
+   * Internal method to get {@link FileSystem} for {@code spec}.
+   */
+  @VisibleForTesting
+  public static FileSystem getFileSystemInternal(URI uri) {
+    String scheme =
+        (uri.getScheme() != null ? uri.getScheme() : LocalFileSystemRegistrar.LOCAL_FILE_SCHEME);
+    return getRegistrarInternal(scheme).fromOptions(REGISTRAR_DEFAULT_CONFIG_MAP.get(scheme));
+  }
+
+  /**
+   * Internal method to get {@link FileSystemRegistrar} for {@code scheme}.
+   */
+  @VisibleForTesting
+  static FileSystemRegistrar getRegistrarInternal(String scheme) {
+    if (REGISTRAR_MAP.containsKey(scheme)) {
+      return REGISTRAR_MAP.get(scheme);
+    } else if (REGISTRAR_MAP.containsKey(DEFAULT_SCHEME)) {
+      return REGISTRAR_MAP.get(DEFAULT_SCHEME);
+    } else {
+      throw new IllegalStateException("Unable to find handler for " + scheme);
+    }
+  }
+
+  /**
+   * Adds the {@link FileSystemRegistrar} to the static {@code REGISTRAR_MAP}.
+   *
+   * @throws IllegalStateException if multiple {@link FileSystem FileSystems}
+   * for the same scheme are detected.
+   */
+  @VisibleForTesting
+  private static void addFileSystemRegistrar(FileSystemRegistrar registrar) {
```

Comment:
This method is pointless since checkDuplicateScheme will catch all failures, might as well inline the REGISTRAR_MAP.put() into loadFileSystemRegistrars()